clang/test/CodeGenCXX/debug-info-line-if-2.cpp

```diff
@@ -0,0 +1,45 @@
+// RUN: %clang_cc1 -debug-info-kind=limited -gno-column-info -triple=x86_64-pc-linux -emit-llvm %s -o - | FileCheck  %s
+
+// The important thing is that the compare and the conditional branch have
+// locs with the same scope (the lexical block for the 'if'). By turning off
+// column info, they end up with the same !dbg record, which halves the number
+// of checks to verify the scope.
+
+int c = 2;
+
+int f() {
+#line 100
+  if (int a = 5; a > c)
+    return 1;
+  return 0;
+}
+// CHECK-LABEL: define {{.*}} @_Z1fv()
+// CHECK:       = icmp {{.*}} !dbg [[F_CMP:![0-9]+]]
+// CHECK-NEXT:  br i1 {{.*}} !dbg [[F_CMP]]
```

Comment:
How come this case ^ already works (so far as I can tell) without this patch, but the others don't? Because we introduce a scope for the variable declaration in a way that we don't in the (cond) case?

Reply:
Ah, yes. That's because EmitIfStmt calls EmitStmt on the initialization statement, which calls EmitStopPoint, which creates a DILocation with the correct scope; this implicitly applies to the rest of the 'if' statement. For (int a = foo()) case, it just issues a declaration for the variable and then proceeds with a normal expression evaluation; the declaration doesn't call EmitStopPoint.

Reply:
Hmm, yeah, that `EmitStopPoint` seems a bit unstable/unreliable - the scoped location handling is designed to be more robust to ensure locations don't "leak out" beyond where they're meant to apply... 

I think maybe `EmitStopPoint` should be removed/reconsidered, but that's perhaps beyond the scope (har har) of this issue - but thoughts in case anyone else feels like picking up and running with that.

How's this location compare to other control structures (loops, etc) - do we (& GCC) use the condition as the location for the branch instructions, or would it be more suitable to use the start of `if` itself?

Reply:
& I guess this all only applies when the control structure doesn't have `{}`? (that being why the test doesn't have braces?)

Reply:
I followed this where my test case led me. You're right it's worth looking at other control structures that have implied lexical blocks.

> & I guess this all only applies when the control structure doesn't have `{}`? (that being why the test doesn't have braces?)

No, this applies in all cases.
```
if (int a = 5; b > c) {
    return a;
} else {
    return a + 1;
}
```
The implied lexical block started by the `if` statement spans its then and else parts; whether those parts are simple or compound doesn't matter. Analogous to how a function definition has an implied lexical block at the opening paren of the parameter list; the function name is visible in the containing scope, the parameter names are not.

Reply:
Re. other statements: `while` does not appear to have a problem, however `for` does. Looking at the implementation, I would have expected it to crop up only when the first `for` clause is empty, but in fact it happens regardless. It conjures up not one but two DILexicalBlocks, both pointing to the `for` keyword. I'd prefer to look into that more deeply as a separate task.